src/wallet/wallet_tx_builder.cpp

```diff
@@ -0,0 +1,751 @@
+
+    // Determine the account we're sending from.
+    auto sendFromAccount = wallet.FindAccountForSelector(selector).value_or(ZCASH_LEGACY_ACCOUNT);
+
+    // We do not set a change address if there is no change.
+    std::optional<ChangeAddress> changeAddr;
+    auto changeAmount = spendable.Total() - resolvedPayments.Total() - fee;
+    if (changeAmount > 0) {
+        auto allowedChangeTypes = [&](const std::set<ReceiverType>& receiverTypes) -> std::set<OutputPool> {
+            std::set<OutputPool> result{resolvedPayments.GetRecipientPools()};
+            // We always allow shielded change when not sending from the legacy account.
+            if (sendFromAccount != ZCASH_LEGACY_ACCOUNT) {
+                result.insert(OutputPool::Sapling);
+            }
+            for (ReceiverType rtype : receiverTypes) {
+                switch (rtype) {
+                    case ReceiverType::P2PKH:
+                    case ReceiverType::P2SH:
+                        // TODO: This is the correct policy, but it’s a breaking change from
+                        //       previous behavior, so enable it separately.
+                        // if (strategy.AllowRevealedRecipients()) {
+                        if (!spendable.utxos.empty() || strategy.AllowRevealedRecipients()) {
+                            result.insert(OutputPool::Transparent);
+                        }
+                        break;
+                    case ReceiverType::Sapling:
+                        if (!spendable.saplingNoteEntries.empty() || strategy.AllowRevealedAmounts()) {
+                            result.insert(OutputPool::Sapling);
+                        }
+                        break;
+                    case ReceiverType::Orchard:
+                        if (params.GetConsensus().NetworkUpgradeActive(anchorHeight, Consensus::UPGRADE_NU5)
+                            && (!spendable.orchardNoteMetadata.empty() || strategy.AllowRevealedAmounts())) {
+                            result.insert(OutputPool::Orchard);
+                        }
+                        break;
+                }
+            }
+            return result;
+        };
+
+        std::visit(match {
+            [&](const CKeyID& keyId) {
+                auto sendTo = pwalletMain->GenerateChangeAddressForAccount(
+                        sendFromAccount,
+                        allowedChangeTypes({ReceiverType::P2PKH}));
+                if (sendTo.has_value()) {
+                    resolvedPayments.AddPayment(
+                            ResolvedPayment(std::nullopt, sendTo.value(), changeAmount, std::nullopt, true));
+                    changeAddr = sendTo.value();
+                }
+            },
+            [&](const CScriptID& scriptId) {
+                auto sendTo = pwalletMain->GenerateChangeAddressForAccount(
+                        sendFromAccount,
+                        allowedChangeTypes({ReceiverType::P2SH}));
+                assert(sendTo.has_value());
```

Comment:
Good question. I just went back to check your original code – there everything is an assertion, except for `UnifiedAddress` and `UnifiedFullViewingKey`. I changed it to match that, and the tests pass (which makes sense – if it _does_ get through without asserting, we would just hit `AddressResolutionError::ChangeAddressSelectionError`.

But then do you know why it would use `if` in the `Unified*` cases?

Reply:
Even eliminating _those_ `if`s, there’s also
```c++
auto zufvk = pwalletMain->GetUFVKForAddress(ua);
if (zufvk.has_value()) {
```
in the `UnifiedAddress` case, so
```c++
// this should be unreachable, but we handle it defensively
```
on `ChangeAddressSelectionError` is misleading.

Reply:
These all `assert` now.